proposals/esm-integration/README.md

```diff
@@ -1,3 +1,308 @@
+These graphs are expressed with import and export statements.
+
+```
+// main.js
+
+import {count, incrementCount} from "./dep.js"
+```
+
+```
+//dep.js
+
+let count = 10;
+
+function incrementCount() {
+    count++;
+}
+
+export {count, incrementCount};
+```
+
+WebAssembly modules currently can't take part in these graphs. This means that developers who want to use WebAssembly modules have to manually trigger instantiation, which may not have completed by the time the JS module graph returns, breaking the mental model of modules.
+
+## Proposed solution
+
+Both goals can be addressed by enabling WebAssembly modules to be loaded using the ES module system. 
+
+This system has three phases:
+
+1. Construction — Module files are fetched and parsed and the module record is constructed.
+1. Instantiation — Exports and imports are wired up to memory locations. Functions are also initialized here.
+1. Evaluation — Top-level module code is evaluated so that non-function exports can be assigned their values.
+
+The work completed in these phases is defined in two different specs:
+
+- A host-specific specification (i.e. the HTML spec) defines how to load modules.
+- The ES module spec provides methods for loaders to use to parse, instantiate, and evaluate modules. The spec defines both an Abstract Module Record and a concrete subclass of this for JS modules, called the Source Text Module Record.
+
+This proposal will create a new WebAssembly Module subclass of the Abstract Module Record. For this, the 3 phases will be specified as follows (with the possible introduction of a sub-phase in Instantiation):
+
+### Construction
+
+During construction, all files for modules in the graph are downloaded and are parsed into Module Records. 
+
+Three things happen for each module during the Construction phase.
+
+1. Finding the file — Based on the module specifier, the host figures out where to get the file from using its module resolution algorithm. Module resolution algorithms differ between hosts. There is ongoing work to fix incompatibilities between Node.js and browser module resolution with the [package-name-maps](https://github.com/domenic/package-name-maps) proposal.
+1. Loading the file — The host fetches the file using the module path generated in module resolution. For browsers, the loading algorithm is specified in the [HTML spec](https://html.spec.whatwg.org/#fetch-a-module-script-tree). The WebAssembly MIME type will need to be added to this spec, but loading should largely work without change for WebAssembly modules. 
+1. Parsing the file — The runtime parses the file into a Module Record. For JavaScript, this work is specified in the [`ParseModule` method](https://tc39.github.io/ecma262/#sec-parsemodule) of the Source Text Module Record. This proposal will add a `ParseModule` method to the WebAssembly spec. The runtime creates the logical equivalent of `WebAssembly.Module` during this phase, and fills in the `RequestedModules` field of the module.
```

Comment:
Isn't `ParseModule` effectively equivalent to `validate`? What's saved off isn't observable I think, so it's a quality of implementation issue whether parsed result is re-parsed after this phase?

Reply:
I'm not sure if I'm fully understanding, so let me know if I misunderstood.

`ParseModule` returns a module record. The [loading algorithm](https://html.spec.whatwg.org/#fetch-the-descendants-of-a-module-script) in the HTML spec then looks for the `RequestedModules` property on that returned Module Record. I don't think that `validate` currently does this (and we probably don't want it to). Calling `validate` should likely be part of the `ParseModule` operation though.

Reply:
You might be right. My thinking is that the module record isn't observable, and all you know is whether you can get one (because it'll validate) or whether you can't. So whether validate creates the module record or not seems irrelevant because you can't observe that, all you can see is whether it validated.

Put another way: I think an implementation could delay the creation of the module record.